cilium: Add option to masq to source route

borkmann opened the pull request:
See commit msgs.

borkmann committed:
cilium, iptables: Support option to masquerade source IP from routing layer

Add a new `--enable-masquerade-to-route-source` option to the agent which
supports more fine-grained masquerading for advanced use cases. For example,
a routing agent installs several CIDRs for a given interface and each of
the CIDRs have a different source address selection criteria. Moreover, the
next hops could even be part of a next hop group where the underlying next
hops are reachable via different interfaces which may not be known to Cilium.

Anyway, if the option is enabled, then the regular MASQUERADE target will
only act as a catchall/fallback, and we install several j SNAT --to-source
rules before that based on the routing setup (destination CIDR & source IP
to use), so that Pod traffic can be masqueraded differenty. For the default
route we keep relying on MASQUERADE target. Note that the user needs to
ensure that there are no overlapping destination prefixes since iptables is
only processing its rules linearly as opposed to routing layer.

Example:

  --enable-masquerade-to-route-source=false (default):
  [...]
  [0:0] -A CILIUM_POST_nat -s 10.244.0.0/24 -m set --match-set cilium_node_set_v4 dst -m comment --comment "exclude traffic to cluster nodes from masquerade" -j ACCEPT
  [0:0] -A CILIUM_POST_nat -s 10.244.0.0/24 ! -d 10.244.0.0/16 ! -o cilium_+ -m comment --comment "cilium masquerade non-cluster" -j MASQUERADE
  [...]

  --enable-masquerade-to-route-source=true:
  [...]
  [0:0] -A CILIUM_POST_nat -s 10.244.0.0/24 -m set --match-set cilium_node_set_v4 dst -m comment --comment "exclude traffic to cluster nodes from masquerade" -j ACCEPT
  [0:0] -A CILIUM_POST_nat -s 10.244.0.0/24 -d 1.1.1.1/32 -m comment --comment "cilium snat non-cluster via source route" -j SNAT --to-source 192.168.2.99
  [0:0] -A CILIUM_POST_nat -s 10.244.0.0/24 -d 192.168.2.13/32 -o enp5s0 -m comment --comment "cilium snat non-cluster via source route" -j SNAT --to-source 192.168.2.99
  [0:0] -A CILIUM_POST_nat -s 10.244.0.0/24 ! -d 10.244.0.0/16 ! -o cilium_+ -m comment --comment "cilium masquerade non-cluster" -j MASQUERADE
  [...]

In the latter, for the destination of 1.1.1.1/32 we SNAT to 192.168.2.99 instead of
192.168.2.12 which is the interface primary address. Similar for 192.168.2.13/32
which also has the -o enp5s0 as device:

  # ip r
  [...]
  1.1.1.1 nhid 40 src 192.168.2.99
  	nexthop via 192.168.2.1 dev enp5s0 weight 1
  	nexthop via 192.168.2.2 dev enp5s0 weight 1
  192.168.2.13 dev enp5s0 scope link src 192.168.2.99
  [...]

Support has been implemented for IPv4 and IPv6 masquerading. BPF-based masquerading
is out of scope for now as we first need to land fib lookup extensions into the
kernel (@brb).

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, iptables: Do not hard-code ip6tables string

Don't hardcode, but check based on prog as we do elsewhere.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

kaworu reviewed (approved):
Helm changes LGTM.

brb commented on Documentation/network/concepts/masquerading.rst:
```diff
@@ -166,3 +166,11 @@ option ``egress-masquerade-interfaces: eth0`` can be used.
+primary interface address. The latter is then only considered as a catch-all
+fallback, and for the default routes. For these advanced cases the user needs
+to ensure that there are no overlapping destination CIDRs as routes on the
+relevant masquerading interfaces.
```
Nit: It'd be nice to add example (from the commit msg) to illustrate the case.

brb commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1217,7 +1219,77 @@ func (m *IptablesManager) installMasqueradeRules(prog iptablesInterface, ifName,
+	// MASQUERADE further below.
+	if option.Config.EnableMasqueradeRouteSource {
+		devices := option.Config.GetDevices()
+		if option.Config.EgressMasqueradeInterfaces != "" {
```
Maybe we could change the `EgressMasqueradeInterfaces` flag type to `StringSlice`? Then the split below would not be needed.

borkmann replied:
Makes sense, in prior cases we just use the plain EgressMasqueradeInterfaces string, but I guess it makes sense to switch to slice and have it the other way around.

borkmann replied:
Ok, done.

borkmann replied:
Looks like I'm running into https://github.com/cilium/cilium-cli/issues/1896 where CI breaks given cilium-cli vendors in cilium itself :/

borkmann replied:
Tobias gave some hints how to work-around this by keeping the old EnableMasqueradeRouteSource alive as string.

brb commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1217,7 +1219,77 @@ func (m *IptablesManager) installMasqueradeRules(prog iptablesInterface, ifName,
+		if prog == ip6tables {
+			family = netlink.FAMILY_V6
+		}
+		if option.Config.EnableIPv4 && prog == ip4tables ||
```
I think we don't need to check for `EnableIPv{4,6}` here, as the callers of the `installMasqueradeRules()` already do the check.

borkmann replied:
good point, agree. fixed

brb commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1217,7 +1219,77 @@ func (m *IptablesManager) installMasqueradeRules(prog iptablesInterface, ifName,
+	// If this option is enabled, then it takes precedence over the catch-all
+	// MASQUERADE further below.
+	if option.Config.EnableMasqueradeRouteSource {
+		devices := option.Config.GetDevices()
```
Do we actually want to fallback to `--devices` if `--egress-masquerade-interfaces` is not set? Asking, as AFAIK `--devices` are only used to control the NodePort BPF and the BPF masq, while previously  the iptables-based MASQ didn't consider them.

borkmann replied:
The source selection can be tied to devices unless next hop groups are used, so both yes & no with regards to their need. I could bail out with error if r.LinkIndex > 0 and len(devices) is 0.

borkmann replied:
Implemented the above now.

brb reviewed (commented):
Nice work! A few comments.

giorio94 reviewed (approved):
sig-agent changes LGTM

borkmann committed:
cilium, iptables: Convert EgressMasqueradeInterfaces to string slice

Convert EgressMasqueradeInterfaces to string slice so that it can be easily
iterated rather than first having to call into ...

  strings.Split(option.Config.EgressMasqueradeInterfaces, ",")

... to get to the array. For the existing cases this means, we need to join
the array instead.

Suggested-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
